staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,342 @@
+		close(closedCh)
+		return closeHandler(code, text)
+	})
+	c.connMu.Unlock()
+	t := time.NewTicker(period)
+	defer t.Stop()
+	for {
+		select {
+		case <-closedCh:
+			return
+		case <-t.C:
+			c.connMu.Lock()
+			if err := c.conn.WriteControl(gwebsocket.PingMessage, nil, time.Now().Add(period)); err != nil {
+				klog.V(7).Infof("Websocket Ping failed: %v", err)
+				// Continue, in case this is a transient failure.
+				// c.conn.CloseChan above will tell us when the connection is
+				// actually closed.
+			} else {
+				klog.V(7).Infof("Websocket Ping succeeeded")
+			}
+			c.connMu.Unlock()
+		}
+	}
+}
+
+func (c *wsStreamCreator) run() {
+	readBuffer := make([]byte, remotecommand.WebSocketBufferSize)
+	// Set up handler for ping/pong heartbeat.
+	if err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline)); err != nil {
+		klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+	}
+	c.conn.SetPongHandler(func(string) error {
+		err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline))
+		if err != nil {
+			klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+		}
+		return nil
+	})
+	for {
+		messageType, r, err := c.conn.NextReader()
+		if err != nil {
+			websocketErr, ok := err.(*gwebsocket.CloseError)
+			if ok && websocketErr.Code == gwebsocket.CloseNormalClosure {
+				err = nil // readers will get io.EOF as it's a normal closure
+			} else {
+				err = fmt.Errorf("next reader: %w", err)
+			}
+			c.closeAllStreamReaders(err)
+			return
+		}
+		if messageType != gwebsocket.BinaryMessage {
+			c.closeAllStreamReaders(fmt.Errorf("unexpected message type: %d", messageType))
+			return
+		}
+		// it's ok to read just a single byte because the underlying library wraps the actual connection with
+		// a buffered reader anyway.
+		_, err = io.ReadFull(r, readBuffer[:1])
+		if err != nil {
+			c.closeAllStreamReaders(fmt.Errorf("read stream id: %w", err))
+			return
+		}
+		streamID := readBuffer[0]
+		s := c.getStream(streamID)
+		if s == nil {
+			klog.Errorf("Unknown stream id %d, discarding message", streamID)
+			continue
+		}
+		for {
+			nr, errRead := r.Read(readBuffer)
+			if nr > 0 {
+				_, errWrite := s.writePipe.Write(readBuffer[:nr])
+				if errWrite != nil {
+					// Pipe must have been closed by the stream user. Nothing to do, discard the message.
+					break
+				}
+			}
+			if errRead != nil {
+				if errRead == io.EOF {
+					break
+				}
+				c.closeAllStreamReaders(fmt.Errorf("read message: %w", err))
+				return
+			}
+		}
+	}
+}
+
+// closeAllStreamReaders closes readers in all streams.
+// This unblocks all stream.Read() calls.
+func (c *wsStreamCreator) closeAllStreamReaders(err error) {
+	for _, s := range c.streams {
+		// Closing writePipe unblocks all readPipe.Read() callers and prevents any future writes.
+		_ = s.writePipe.CloseWithError(err)
+	}
+}
+
+type stream struct {
+	headers   http.Header
+	readPipe  *io.PipeReader
+	writePipe *io.PipeWriter
```

Comment:
I'm not sure a io.Pipe is the right thing to do here, since it blocks writes until the reader finish https://pkg.go.dev/io#Pipe, and since we are holding the connection lock, until the stream reader reads from the pipe new Writes on the pipe will be blocked.

If we use a buffer https://pkg.go.dev/bytes@go1.20.6#NewBuffer, the connection handler will process the messages and write each stream data to the corresponding buffer, and the readers can read from it, since the buffer is not thread safe we had to handle the concurrency with a new lock in the stream

WDYT?

Reply:
The connection lock is not being held during the pipe write. A pipe for each stream is used to de-multiplex data from the websocket `Connection`. The connection lock is not used during this de-multiplexing loop. The data read from the `Connection` is read into a buffer before it is written to the stream pipe. If the pipe write blocks for this buffered data from the `Connection`, it is because the data hasn't been adequately consumed by the stream pipe reader. Even if we build our own stream buffers with our own concurrency, we would run into the same problem. I don't think adding complexity by building our own concurrent stream buffers gives us anything.

Reply:
did this discussion get resolved and is the resolution well-tested?

Reply:
I am confident this is a non-issue. Antonio left on vacation before we could resolve it.

Reply:
+1